### ISSUE TITLE
[BUG] Investigate failed e2e virtual machine function test cases for v1.4.0

### ISSUE THREAD

**Issue by TachunLin** [harvester/harvester#6599]
**Describe the bug**

According to the latest e2e automation test [harvester_run_test/107/TestReport](http://172.19.99.74/job/harvester-runtests/107/TestReport/). 
The test focus on virtual machine function tests ([test_3_vm_functions](https://github.com/harvester/tests/blob/main/harvester_e2e_tests/integrations/test_3_vm_functions.py))

We found the following failed test cases: 

- TestVMOperations::test_stop
- TestVMOperations::test_restart
- TestVMOperations::test_restart
- TestVMOperations::test_softreboot
- TestVMOperations::test_migrate
- TestVMOperations::test_abort_migrate
- TestVMResource::test_create_schedule_on_maximum[cpu]
- TestVMResource::test_create_schedule_on_maximum[memory]
- TestVMResource::test_update_schedule_on_maximum[cpu]
- test_update_schedule_on_maximum[memory]
- TestVMResource::test_update_enable_usb
- TestVMResource::test_update_enable_user_data
- TestVMClone::test_clone_stopped_vm
- TestVMWithVolumes::test_create_with_two_volumes
- test_vm_with_bogus_vlan 

  ![image](https://github.com/user-attachments/assets/c22fd773-ca3f-48c4-a180-3d927cbc7556)
  ![image](https://github.com/user-attachments/assets/db6599c0-1442-4799-bf79-c3cf3649f488)

* Thus we create this issue to investigate these failed vm function e2e tests to know if there are new issue introduced from `v1.4.0` 

**To Reproduce**
1. Clone the harvester/tests repository 
2. Edit the config.yml and set appropriate setting
3. Execute the backend e2e test on the `test_3_vm_functions.py` tests only 
4. Check the test result 
5. Also can use the harvester-run-test Jenkins job.

**Expected behavior**

* Check if any potential issue introduced from v1.4.0, if so, create a new issue to track.  
* Fix the rest of the e2e test cases coming from feature changes or bug fixed 

**Environment**
 - Harvester ISO version: `v1.4-cd10d18-head` (24/09/19) 
 - Underlying Infrastructure: Single node bare metal machine

**Comment by lanfon72** [comment]
I think this should be created in tests repo rather than harvester

**Comment by TachunLin** [comment]
Thanks for the suggestion. 
I just create the related issue https://github.com/harvester/tests/issues/1540 in test repo to track 

This issue will be closed afterward once after confirmation.

**Comment by lanfon72** [comment]
you can just use the `Transfer issue` to move this to *tests* repo, it would be more convienient.
![image](https://github.com/user-attachments/assets/b4b42e3a-f87c-4d29-8ce2-1b4e0da3dd09)

**Comment by khushboo-rancher** [comment]
@TachunLin Are these cases failing when tried manually also?

**Comment by TachunLin** [comment]
### Analyzing failed vm function test with manual execution

Use manual steps to check the failed vm function tests result and rerun the e2e test:

<details><summary>$\color{green}{\textsf{PASS}}$ Manual check and analyze: "test_stop" vm function $~~$</summary>

1. Monitoring the e2e execution, we can find the vm have already been stopped 
![image](https://github.com/user-attachments/assets/69b459c0-85f4-48da-9818-a53a0584e665)

2. Further check the follow up status assertion in 
  ```
          assert "Halted" == data['spec']['runStrategy']
          assert "Stopped" == data['status']['printableStatus']
  ```
3. And to compare with the actual status, we can find the 
  ```
  (Pdb) (data['spec']['runStrategy']
  Pdb) 'RerunOnFailure'
  data['status']['printableStatus']
  (Pdb) 'Stopped'
  ``` 
4. It may relate to the changes https://github.com/harvester/harvester/issues/6172#issuecomment-2362908672

 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Manual check and analyze: "test_restart" vm function $~~$</summary>

1. Use manual steps, we can correctly restart the vm used in the e2e test

2. According to the failure log, we may assume the failure may caused by previous test failure. 

  - At test_3_vm_functions: Line 355
  ```
  E       AssertionError: unable to get VM(vm-09h47m05s099439-09-19)'s instance infos with errors:
  E         Status(404): {'type': 'error', 'links': {}, 'code': 'NotFound', 'message': 'virtualmachineinstances.kubevirt.io "vm-09h47m05s099439-09-19" not found', 'status': 404}
  E       assert 200 == 404
  ```

 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Manual check and analyze: "test_softreboot" vm function $~~$</summary>

1. Use manual steps, we can correctly softreboot the vm

2. According to the failure log, we may assume the failure may caused by previous test failure. 

  - At test_3_vm_functions: Line 402
  ```
  E       AssertionError: unable to get VM(vm-09h47m05s099439-09-19)'s instance infos with errors:
  E         Status(404): {'type': 'error', 'links': {}, 'code': 'NotFound', 'message': 'virtualmachineinstances.kubevirt.io "vm-09h47m05s099439-09-19" not found', 'status': 404}
  E       assert 200 == 404
  ```

3. When we bypass the `test stop`, the test_softreboot` can execute correctly 
  ```
  harvester_e2e_tests/integrations/test_3_vm_functions.py::test_minimal_vm PASSED [  3%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMOperations::test_pause PASSED [  6%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMOperations::test_unpause PASSED [  9%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMOperations::test_start SKIPPED [ 12%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMOperations::test_restart PASSED [ 15%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMOperations::test_softreboot PASSED [ 18%]
  
  ```
4. Thus we can assume the failure was coming from the previous test execution not well completed

 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Manual check and analyze: "test_update_enable_usb" vm function $~~$</summary>

1. Use manual steps can correctly enable the "Enable USB Tablet" and start vm well

2. According to the failure log, we may assume the failure may caused by previous test failure. 

  - At test_3_vm_functions: Line 855

  ```
  >       assert 200 == code, (code, data)
  E       AssertionError: (409, {'code': 'Conflict', 'links': {}, 'message': 'Operation cannot be fulfilled on virtualmachines.kubevirt.io "vm-0...19": the object has been modified; please apply your changes to the latest version and try again', 'status': 409, ...})
  E       assert 200 == 409
  ```

3. When we bypass the `test stop`, the test_update_enable_usb` can execute correctly 

  ```
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMResource::test_update_cpu PASSED [ 46%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMResource::test_update_enable_usb PASSED [ 50%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMResource::test_update_enable_user_data PASSED [ 53%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMClone::test_clone_running_vm PASSED [ 56%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMClone::test_clone_stopped_vm FAILED [ 59%]
  
  ```

4. Thus we can assume the failure was coming from the previous test execution not well completed

 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Manual check and analyze: "test_update_enable_user_data" vm function $~~$</summary>

1. Use manual steps can correctly enable the "Install guest agent" and start vm well with guest-qemu service enabled

2. According to the failure log, we may assume the failure may caused by previous test failure. 

  - At test_3_vm_functions: Line 884
  
  ```
          # Create a stopped VM without user data (guest agent)
          code, data = api_client.vms.create(unique_vm_name, vm_spec)
  >       assert 201 == code, (code, data)
  E       AssertionError: (409, {'code': 'AlreadyExists', 'links': {}, 'message': 'virtualmachines.kubevirt.io "vm-09h47m05s099439-09-19" already exists', 'status': 409, ...})
  E       assert 201 == 409
  ```


3. When we bypass the `test stop`, the `test_update_enable_user_data` can execute correctly 

  ```
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMResource::test_update_cpu PASSED [ 46%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMResource::test_update_enable_usb PASSED [ 50%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMResource::test_update_enable_user_data PASSED [ 53%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMClone::test_clone_running_vm PASSED [ 56%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMClone::test_clone_stopped_vm FAILED [ 59%]
  ```

4. Thus we can assume the failure was coming from the previous test execution not well completed

 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ Manual check and analyze: "test_create_with_two_volumes" vm function $~~$</summary>

1. Use manual steps to create vm with two volumes, check can start vm well. 

2. According to the failure log, we may assume the failure may caused by previous test failure. 

  - At test_3_vm_functions: Line 1275
  
  ```
          # Start VM with 2 additional volumes
          code, data = api_client.vms.update(unique_vm_name, vm_spec)
  >       assert 200 == code, (code, data)
  E       AssertionError: (409, {'code': 'Conflict', 'links': {}, 'message': 'Operation cannot be fulfilled on virtualmachines.kubevirt.io "stop...19": the object has been modified; please apply your changes to the latest version and try again', 'status': 409, ...})
  E       assert 200 == 409
  ```


3. When we bypass the `test stop`, the `test_create_with_two_volumes` can execute correctly 

  ```
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMClone::test_clone_running_vm PASSED [ 56%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMClone::test_clone_stopped_vm FAILED [ 59%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMWithVolumes::test_create_with_two_volumes PASSED [ 62%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMWithVolumes::test_create_with_existing_volume PASSED [ 65%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMWithVolumes::test_create_with_volume_image PASSED [ 68%]
  ```


4. Thus we can assume the failure was coming from the previous test execution not well completed

 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Manual check and analyze: "test_vm_with_bogus_vlan" vm function $~~$</summary>

1. According to the failure log, we may assume the failure may caused by previous test failure. 

  - At test_3_vm_functions: Line 1835
  
  ```
          code, data = api_client.vms.create(unique_vm_name, vm)
      
  >       assert 201 == code, (code, data)
  E       AssertionError: (409, {'code': 'AlreadyExists', 'links': {}, 'message': 'virtualmachines.kubevirt.io "vm-09h47m05s099439-09-19" already exists', 'status': 409, ...})
  E       assert 201 == 409
  ```


2. When we bypass the `test stop`, the `test_update_enable_user_data` can execute correctly 

  ```
  harvester_e2e_tests/integrations/test_3_vm_functions.py::test_update_vm_machine_type[q35_to_pc] SKIPPED437) [ 87%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::test_update_vm_machine_type[pc_to_q35] SKIPPED437) [ 90%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::test_vm_with_bogus_vlan PASSED [ 93%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestHotPlugVolume::test_add PASSED [ 96%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestHotPlugVolume::test_remove PASSED [100%]
  
  ```

3. Thus we can assume the failure was coming from the previous test execution not well completed

 </details>


<details><summary>$\color{purple}{\textsf{Need discussion}}$ Manual check and analyze: "test_clone_stopped_vm" vm function $~~$</summary>

1. Use manual steps to create a vm, write some data, stop the vm and clone to vm. The vm can be cloned and running well with data exists
![image](https://github.com/user-attachments/assets/0100c6ea-c997-472f-8c50-15bd167ab0d4)

2. According to the failure log, we may assume the failure may caused by previous test failure. 

  - At test_3_vm_functions: Line 1170
  
  ```
  >           raise AssertionError(
                  f"Cloned VM {cloned_name} is not available and stopped"
                  f"Status({code}): {data}"
              )
  E           AssertionError: Cloned VM cloned-stopped-0929589235-vm-09h47m05s099439-09-19 is not available and stoppedStatus(200): {'id': 'default/cloned-stopped-0929589235-vm-09h47m05s099439-09-19', 'type': 'kubevirt.io.virtualmachine', 'links': {'remove': 'https://172.19.108.29/v1/harvester/kubevirt.io.virtualmachines/default/cloned-stopped-0929589235-vm-09h47m05s099439-09-19', 'self': 
  ```

2. Further check the follow up status assertion in 

  ```
                 and "Halted" == data['spec'].get('runStrategy')
                 and "Stopped" == data.get('status', {}).get('printableStatus')):
  ```

3. And to compare with the actual status, we can find the cloned vm is in `Running` state

  ![image](https://github.com/user-attachments/assets/2630d57a-a998-4905-aac6-71629b438216)

4. And the vm runStrategy is `Halted`
    ```
    > /home/davidtclin/Documents/Harvester/Project_repo/e2e_test/tests/harvester_e2e_tests/integrations/test_3_vm_functions.py(1163)test_clone_stopped_vm()
    -> endtime = datetime.now() + timedelta(seconds=wait_timeout)
    code, data = api_client.vms.get(cloned_name)
    (Pdb) (data['spec'].get('runStrategy')
    Pdb) 'RerunOnFailure'
    data.get('status', {}).get('printableStatus')
    (Pdb) 'Running'
    ```

5. It may relate to the changes https://github.com/harvester/harvester/issues/6172#issuecomment-2362908672

6. And we also need to confirm whether the cloned vm should be in the `Stopped` or `Running` satate

 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ Manual check and analyze: "test_migrate" vm function $~~$</summary>

1. Use manual steps can correctly migrate the vm to other host

2. Retest on multiple nodes Harvester, it can execute well

3. According to the failure log, we may assume the failure may caused by previous test failure. 

  - At test_3_vm_functions: Line 452
  ```
          code, host_data = api_client.hosts.get()
          assert 200 == code, (code, host_data)
          code, data = api_client.vms.get_status(unique_vm_name)
  >       cur_host = data['status'].get('nodeName')
  E       AttributeError: 'int' object has no attribute 'get'
  ```


4. When we bypass the `test stop`, the `test_update_enable_user_data` can execute correctly 

  ```
  harvester_e2e_tests/integrations/test_3_vm_functions.py::test_minimal_vm PASSED [  3%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMOperations::test_pause PASSED [  7%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMOperations::test_unpause PASSED [ 10%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMOperations::test_start SKIPPED [ 14%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMOperations::test_restart PASSED [ 17%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMOperations::test_softreboot PASSED [ 21%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMOperations::test_migrate PASSED [ 25%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMOperations::test_abort_migrate PASSED [ 28%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMOperations::test_delete PASSED [ 32%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::test_create_stopped_vm PASSED [ 35%]
  
  ```

5. Thus we can assume the test would be PASS on the multiple nodes Harvester test environment

 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ Manual check and analyze: "test_abort_migrate" vm function $~~$</summary>

1. Use manual steps can correctly abort the vm migration . 

2. Retest on multiple nodes Harvester, this test can execute well

3. According to the failure log, we may assume the failure may caused by previous test failure. 

  - At test_3_vm_functions: Line 491
  ```
          code, host_data = api_client.hosts.get()
          assert 200 == code, (code, host_data)
          code, data = api_client.vms.get_status(unique_vm_name)
  >       cur_host = data['status'].get('nodeName')
  E       AttributeError: 'int' object has no attribute 'get'
  ```


4. When we bypass the `test stop`, the `test_update_enable_user_data` can execute correctly 

  ```
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMOperations::test_softreboot PASSED [ 21%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMOperations::test_migrate PASSED [ 25%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMOperations::test_abort_migrate PASSED [ 28%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMOperations::test_delete PASSED [ 32%]
  harvester_e2e_tests/integrations/test_3_vm_functions.py::test_create_stopped_vm PASSED [ 35%]
  
  ```

5. Thus we can assume the test would be PASS on the multiple nodes Harvester test environment

 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ test_create_schedule_on_maximum[cpu][memory]" vm function $~~$</summary>


1. Retest on multiple nodes Harvester, this test can execute well

2. According to the failure log, we may assume the failure may caused by previous test failure. 

- At test_3_vm_functions: Line 632
```
        nodes_res = sorted(nodes_res, key=lambda n: n[1][res_type], reverse=True)
>       expected_host, expected_res = next((cn, nd[res_type] + (cd[res_type] - nd[res_type]) / 2)
                                           for (cn, cd), (nn, nd) in zip(nodes_res, nodes_res[1:]))
E       StopIteration
```


3. When we bypass the `test stop`, this test  can execute correctly 

```
harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMOperations::test_migrate PASSED [ 21%]
harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMOperations::test_abort_migrate PASSED [ 25%]
harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMOperations::test_delete PASSED [ 28%]
harvester_e2e_tests/integrations/test_3_vm_functions.py::test_create_stopped_vm PASSED [ 31%]
harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMResource::test_create_schedule_on_maximum[cpu] PASSED [ 34%]
harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMResource::test_create_schedule_on_maximum[memory] PASSED [ 37%]
harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMResource::test_update_schedule_on_maximum[cpu] PASSED [ 40%]
harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMResource::test_update_schedule_on_maximum[memory] PASSED [ 43%]

```


4. Thus we can assume the test would be PASS on the multiple nodes Harvester test environment

 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ "test_update_schedule_on_maximum[cpu][memory] " vm function $~~$</summary>

1. Retest on multiple nodes Harvester, this test can execute well

2. According to the failure log, we may assume the failure may caused by previous test failure. 
  - At test_3_vm_functions: Line 697
  ```
          code, data = api_client.vms.update(unique_vm_name, vm_spec)
  >       assert 200 == code, (code, data)
  E       AssertionError: (409, {'code': 'Conflict', 'links': {}, 'message': 'Operation cannot be fulfilled on virtualmachines.kubevirt.io "stop...19": the object has been modified; please apply your changes to the latest version and try again', 'status': 409, ...})
  E       assert 200 == 409
  ```


3. When we bypass the `test stop`, this test can execute correctly 

```
harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMOperations::test_migrate PASSED [ 21%]
harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMOperations::test_abort_migrate PASSED [ 25%]
harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMOperations::test_delete PASSED [ 28%]
harvester_e2e_tests/integrations/test_3_vm_functions.py::test_create_stopped_vm PASSED [ 31%]
harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMResource::test_create_schedule_on_maximum[cpu] PASSED [ 34%]
harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMResource::test_create_schedule_on_maximum[memory] PASSED [ 37%]
harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMResource::test_update_schedule_on_maximum[cpu] PASSED [ 40%]
harvester_e2e_tests/integrations/test_3_vm_functions.py::TestVMResource::test_update_schedule_on_maximum[memory] PASSED [ 43%]

```

4. Thus we can assume the test would be PASS on the multiple nodes Harvester test environment

 </details>


### Test Information
* Test Environment: Three nodes harvester on kvm machines
* Harvester version: `v1.4-cd10d18-head` (23/09/19)

**Comment by TachunLin** [comment]
@khushboo-rancher,
After manual check test steps and retest all failed vm function e2e test. 
According to the previous check result: 


We found only the `test_stop` and `test_cloned_stopped_vm` have reproducible failure. 

1. The `test_stop` was failed due to the script tried to check the `assert "Halted" == data['spec']['runStrategy']` 
2. While the actual vm `runStrategy` is `'RerunOnFailure'`. The vm have already in stopped state.  
    - Double check with manual steps, we can also stop the vm correctly
    - It may relate to the changes https://github.com/harvester/harvester/issues/6172#issuecomment-2362908672

3. The `test_clone_stopped_vm` was failed since the script tried to check the `"Halted" == data['spec'].get('runStrategy')` and `"Stopped" == data.get('status', {}).get('printableStatus'))` 
4. While the actual vm `runStrategy` is `'RerunOnFailure'` and the vm `status` is `Running`
5. **Thus bring up the discussion for when we cloned a stopped vm, the vm should started in `running` or `keep` in stopped state** 

6. The `test_stop` test failure seems would cause the follow up tests failed too.  
7. When we remove the `test_stop`, the following test can execute correctly 

    - test_restart
    - test_softreboot
    - test_update_enable_usb
    - test_update_enable_user_data
    - test_create_with_two_volumes

8. When we retest the failed cases on multiple nodes Harvester, the following test can execute correctly 

    - test_migrate
    - test_abort_migrate
    - test_create_schedule_on_maximum[cpu][memory]
    - test_update_schedule_on_maximum[cpu][memory]
    
9. On manual execution, most test cases can work as expected. 

In summary, I think no significant issue found on v1.4.0-head for vm function test.  
As per discussed, I would suggest to **transfer** this issue to `tests` repo for further investigation. 
There if found any new issue introduced, we can create a new issue to track.
